full-service/src/db/transaction_log.rs

```diff
@@ -1968,4 +1970,64 @@ mod tests {
 
         assert_eq!(tx_log.tx, expected_tx_log.tx);
     }
+
+    #[test]
+    fn test_try_from_vec_output_txo_for_transaction_id() {
+
+        let mut rng: StdRng = SeedableRng::from_entropy();
+        let root_id = RootIdentity::from_random(&mut rng);
+        let recipient_account_key = AccountKey::from(&root_id);
+
+        let num_txos = 5; // txo vector size
+        let num_loops = 5; // number of times to run test
+        let amount = 77;
+
+        for _ in 1..=num_loops { // run test loop
+
+            let mut output_vec: Vec<OutputTxo> = Vec::new();
+
+            for _ in 0..num_txos { // loop to build vector and test try_from()
+
+                let (tx_out, _) = create_test_txo_for_recipient(
+                            &recipient_account_key,
+                            0, // subaddress_index
+                            Amount::new(amount * MOB, Mob::ID),
+                            &mut rng,
+                        );
+
+                let output_txo = OutputTxo {
+                    tx_out: tx_out.clone(),
+                    recipient_public_address: recipient_account_key.subaddress(0),
+                    confirmation_number: TxOutConfirmationNumber::default(),
+                    amount: Amount::new(amount * MOB, Mob::ID),
+                    shared_secret: None,
+                };
+
+                output_vec.push(output_txo);
+            }
+
+            let min_public_key = output_vec
+                .iter()
+                .min_by_key(|txo| txo.tx_out.public_key)
+                .unwrap()
+                .tx_out
+                .public_key;
+            let transaction_log_id = TransactionId::try_from(output_vec).unwrap();
+
+            assert_eq!(min_public_key.to_string(), transaction_log_id.0);
+        }
+    }
+
+    #[test]
+    fn test_try_from_empty_vec_output_txo_for_transaction_id() {
+
+        let output_vec: Vec<OutputTxo> = Vec::new();
+        let transaction_log_id = TransactionId::try_from(output_vec);
+
+        match transaction_log_id {
+            Ok(_) => assert!(false, "Expected error, but got Ok"),
+            Err(err) => assert_eq!(err, "no valid payload_txo"),
+        }
```

Comment:
I did feel that that Err(err) assertion I had there was a weak point, and raised it in a separate discussion with Sara and Henry.  Last night Henry went deep on the best way implement the match statement or equivalent, and I followed along behind him.  Henry did significantly improve what I had originally.  But what you are showing there, and on a single line no less, outshines us both I think.

@holtzman : I expect to go with what Nick has, but before I resolve this convo, do you have any thoughts?

Reply:
looks great to me.

Reply:
tried out assert_matches! and found that it requires
`#![feature(assert_matches)]`

to be added to the top of the file, as it is considered an `unstable feature`. There seems to be some debate about its semantics.  Further reading here:

https://github.com/rust-lang/rust/issues/82775

Reply:
I was able to get assert_matches to work by adding a `use` statement into the body of the test function:

`use std::assert_matches::assert_matches;`

At the moment I'm not clear if that's resolves to the same version as Henry gets with the `feature` attribute.  Would the version in the `std` crate be considered stable?  If not, should we avoid using this?

Reply:
> tried out assert_matches! and found that it requires
#![feature(assert_matches)] to be added to the top of the file

I believe this should only be required, once, at the crate level in `foo/src/lib.rs` or `foo/src/main.rs`

I've also been using this create at times, https://docs.rs/assert_matches/latest/assert_matches/

I can't decide if the feature gating is a good thing or a bad thing. For instance we have one project where the use of feature gating is the only reason the project is using nightly. The use of nightly is forcing that project to stay back on some dependencies, because those dependencies use nightly for some of their beta work and it causes compilation errors in our usage.

Reply:
Thanks Nick.  I'm going to look at moving the feature into `into full-service/src/lib.rs`.  But I'm still unclear as to whether it is safe to go ahead with this.  I will defer to you and to @holtzman.

Reply:
So it turns out that `#![feature(assert_matches)]` is already in `full-service/src/lib.rs`.  But if I don't declare it somewhere in `transaction_log.rs` then rustc tells me it can't find it.  

Still looking at that right now.  If you want to replicate this, you can comment out the `use` statement I added on line `2018` of `transaction_log.rs` and recompile.

Reply:
I think there maybe some confusion with the `use` versus `#![feature]` nomenclature.

The `#![feature(assert_matches)]` tells the compiler that you know `assert_matches` is unstable and that you still want to work with it. This should only need to be done at the crate level and it's already done for full-service.

the `use std::assert_matches::assert_matches;` brings the `assert_matches` behavior into scope so that it can be used in that test. This is namespacing in case you have a local function or macro called `assert_matches`. The `use` statement needs to be made in every module that one is going to use the `assert_matches!()` macro.
I'd suggest putting this use statement at the top of the test module
```rust
#[cfg(test)]
mod tests {
    use itertools::Itertools;
    use mc_account_keys::{
        AccountKey, PublicAddress, RootIdentity, ShortAddressHash, CHANGE_SUBADDRESS_INDEX,
    };
    ...
    use std::{assert_matches::assert_matches, iter::FromIterator, ops::DerefMut, time::Duration};
```

Reply:
Ah ok.  I did go through that in rust book, but it's clearer now that I'm working through an actual use case.

I happened to be on with Sara and we took a look at the PR together, she let me know that it's not unusual for unstable features to be used.  So I'm going to go ahead and merge.

Reply:
Sorry, update the placement of the `use` statement and then merge.